test/relayserver/RelayServer.test.ts

```diff
@@ -365,18 +366,55 @@ contract('RelayServer', function (accounts) {
   })
 
   describe('#createRelayTransaction()', function () {
-    before(async function () {
-      await env.relayServer.txStoreManager.clearAll()
-    })
-
-    it('should relay transaction', async function () {
+    async function createAndVerifyRelayTransaction (): Promise<RelayTransactionRequest> {
       const req = await env.createRelayHttpRequest()
       assert.equal((await env.relayServer.txStoreManager.getAll()).length, 0)
       await env.relayServer.createRelayTransaction(req)
       const pendingTransactions = await env.relayServer.txStoreManager.getAll()
       assert.equal(pendingTransactions.length, 1)
       assert.equal(pendingTransactions[0].serverAction, ServerAction.RELAY_CALL)
       // TODO: add asserts here!!!
+      return req
+    }
+
+    async function expectRelayTransactionToFail (request: RelayTransactionRequest, msg: string): Promise<void> {
+      try {
+        await env.relayServer.createRelayTransaction(request)
+        fail("relayServer.createRelayTransaction() should have failed, but it didn't")
+      } catch (error) {
+        const errorMsg = (error as Error).message
+        expect(errorMsg).not.to.be.undefined
+        expect(errorMsg).to.include(msg)
+      }
+    }
+
+    beforeEach(async function () {
+      await env.relayServer.txStoreManager.clearAll()
+    })
+
+    it('should relay transaction', async function () {
+      await createAndVerifyRelayTransaction()
+    })
+
+    it('should fail if the same relay transaction is submitted twice', async () => {
+      const relayTxRequest = await createAndVerifyRelayTransaction()
+      await expectRelayTransactionToFail(relayTxRequest, 'Unacceptable relayMaxNonce')
+    })
+
+    describe('with the goal of submitting a relay transaction twice', () => {
+      it('should fail if the relayMaxNonce is manually tampered ', async () => {
+        const relayTxRequest = await createAndVerifyRelayTransaction()
+        relayTxRequest.metadata.relayMaxNonce = relayTxRequest.metadata.relayMaxNonce + 1
+        await expectRelayTransactionToFail(relayTxRequest, 'revert nonce mismatch')
```

Comment:
i am wondering why this doesn't fail with a `revert signature mismatch`, since all of these steps are included in the next test.

is it because it's missing the `relayTxRequest.relayRequest.request.nonce` assignment?

so in a way this would be tampering with the nonce erroneously, right? the next test does it correctly, only the signature doesn't match.

if this is the case, i am ok with it

Reply:
>is it because it's missing the `relayTxRequest.relayRequest.request.nonce` assignment?
yes, indeed:
1. the `relayMaxNonce` is checked by the `RelayServer` -> "Unacceptable relayMaxNonce"
2. if we try to temper it, then the transaction fails because the `SmartWallet` check for the `nonce` -> "nonce mismatch"
3. while if we temper the nonce also, the transaction fails because the `SmartWallet` check for the signature -> "signature mismatch"